stats/stats.go

```diff
@@ -203,24 +206,47 @@ func NewHistogram(Offset float64, Divider float64) *Histogram {
 	return h
 }
 
-// Tradeoff memory for speed (though that also kills the cache so...)
-// this creates an array of 100k (max value) entries
-// TODO: consider using an interval search for the last N big buckets
+// Val2Bucket values are kept in two different structure
+// val2Bucket allows you reach between 0 and 1000 in constant time
 func init() {
-	lastV := int32(lastValue)
-	val2Bucket = make([]int, lastV+1)
+	val2Bucket = make([]int, maxArrayValue)
+	maxArrayValueIndex = -1
+	for i, value := range histogramBucketValues {
+		if value == maxArrayValue {
+			maxArrayValueIndex = i
+			break
+		}
+	}
+	if maxArrayValueIndex == -1 {
+		log.Fatalf("Bug boundary maxArrayValue=%d not found in bucket list %v", maxArrayValue, histogramBucketValues)
+	}
 	idx := 0
-	for i := int32(0); i <= lastV; i++ {
+	for i := int32(0); i < maxArrayValue; i++ {
 		if i >= histogramBucketValues[idx] {
 			idx++
 		}
 		val2Bucket[i] = idx
 	}
-	// coding bug detection (aka impossible if it works once)
-	if idx != numValues {
-		log.Fatalf("Bug in creating histogram buckets idx %d vs numbuckets %d (last val %d)", idx, numValues, lastV)
+	// coding bug detection (aka impossible if it works once) until 1000
+	if idx != maxArrayValueIndex {
+		log.Fatalf("Bug in creating histogram index idx %d vs index %d up to %d", idx, int(maxArrayValue), maxArrayValue)
 	}
+}
 
+// lookUpIdx looks for scaledValue's index in histogramBucketValues
+// TODO: change linear time to O(log(N)) with binary search
+func lookUpIdx(scaledValue int) int {
+	scaledValue32 := int32(scaledValue)
+	if scaledValue32 < maxArrayValue { //constant
```

Comment:
This is not a defensive code. That check allows us to reach lower than 1000 values in the bucket.

Reply:
But we can separate the method if you want. I'm open to suggestions :)

Reply:
Please check TestBucketLookUp method in stats_test.go. I have added some boundary tests for low values as well. If you suggest any values for boundary test, I can add them

Reply:
if you look in the code for record, lookUpIdx() which is private can't be called with values < 1000

Reply:
You mean should I do lookUpIdx public? Even lookUpIdx() is private, record and lookUpIdx is in same file. So I'm expecting that, lookUpIdx method is usable in record. Also inside the lookUpIdx, scaledValue32 < maxArrayValue check allows you to reach values < 1000 as well. If you look the master branch, idx = lookUpIdx(scaledVal) has same responsiblility as idx = val2Bucket[int(scaledVal)]

Reply:
> should I do lookUpIdx public? 

no, no it should stay private, in theory it could just be the end of record() which already private too

you are right and I forgot this replaces 
		idx = val2Bucket[int(scaledVal)]
so it needs to handle the val2Bucket part too; my bad